### PR TITLE
Fix: #23816 Content Type Field Variables Cut Off (Redesign)

### LINKED ISSUE #23816: Content type field variables cut off
[](https://mrkr.io/s/63c1c3d39165a8659ceb96be/0)

## Problem Statement
When the field variable is too long, it cut off.

## Steps to Reproduce
1. Edit a content type
2. Add rows with 4 columns
3. Add one field to each column with a long title

## Acceptance Criteria
- Needs to show Title
- Needs to show Type
- Needs to show Variable Name - with "Copy" icon
- Users have "unlimited" characters... so some truncation is needed.

## Wireframes / Designs / Prototypes
- Needs UI

---

**Reported by:** Freddy Montes (freddy@dotcms.com)

**Source URL:** [http://localhost:8080/dotAdmin/#/content-types-angular/edit/657897dfb36ef211ebfb4128de818787](http://localhost:8080/dotAdmin/#/content-types-angular/edit/657897dfb36ef211ebfb4128de818787)

**Issue details:** [Open in Marker.io](https://app.marker.io/i/63c1c3d39165a8659ceb96c1_36d8a4e2737e5abf?advanced=1)

<table><tr><td><strong>Device type</strong></td><td>desktop</td></tr><tr><td><strong>Browser</strong></td><td>Chrome 109.0.0.0</td></tr><tr><td><strong>Screen Size</strong></td><td>3360 x 1890</td></tr><tr><td><strong>OS</strong></td><td>OS X 13.0.0</td></tr><tr><td><strong>Viewport Size</strong></td><td>1551 x 1805</td></tr><tr><td><strong>Zoom Level</strong></td><td>100%</td></tr><tr><td><strong>Pixel Ratio</strong></td><td>@&#8203;2x</td></tr></table>

```[tasklist]
### Tasks
- [ ] Provide Design @Melissa-dotCMS 
- [ ] Update Feature
```

### PR DESCRIPTION
**Pull request by zJaaal**
### Proposed Changes
* Refactor the styles and layout of `dot-content-type-draggable-item` component
* Small refactor of `dot-copy-button` to be reuse in this design

### Checklist
- [ ] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
image  |  image

#### Another states of this component

image

image

image

### CODE REVIEW

**Comment by zJaaal** [comment]
Although this PR looks done, there's some refactor to be done like:

- [ ] Create a store to maintain the state of the component
- [ ] Ask for a refactor of `dot-icon-button-tooltip` which is being used in several places of the project and it looks outdated taking into account the new designs

**Comment by zJaaal** [comment]
> Although this PR looks done, there's some refactor to be discussed like:
> 
> * [ ]  Ask for a refactor of `dot-icon-button-tooltip` which is being used in several places of the project and it looks outdated in comparison of the new designs

This shouldn't be done in this PR, this would require a lot of effort since it's being used in +20 files

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.html [hunk 1]
```diff
@@ -1,42 +1,36 @@
...
+        <span class="field__name"> {{ field.name }}</span>
+    </div>
+    <div class="field__actions">
+        <p-button
+            id="field__actions-delete"
+            *ngIf="!field.fixed"
+            [pTooltip]="'contenttypes.action.delete' | dm"
+            (click)="removeItem($event)"
+            icon="pi pi-trash"
+            styleClass="p-button-sm p-button-icon-only p-button-transparent"
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.html:14]: Just remove the background in `p-button-icon-only`

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.scss [hunk 2]
```diff
@@ -23,60 +23,89 @@
...
-dot-icon-button-tooltip ::ng-deep .material-icons,
-dot-icon {
-    color: $color-palette-gray-700;
-}
+.field__properties {
+    background: $color-palette-primary-100;
+    padding: $spacing-2 $spacing-3;
+    display: flex;
+    align-items: center;
+    min-height: 73px;
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.scss:31]: Make this ren.

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.scss [hunk 3]
```diff
@@ -23,60 +23,89 @@
...
+            font-size: $font-size-xs;
+            min-height: $font-size-xs;
+        }
 
-.field__name {
-    @include truncate-text;
+        .field__name {
+            @include truncate-text;
+            color: black;
+            font-size: $font-size-md;
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.scss:53]:
```suggestion
            color: $black;
            font-size: $font-size-md;
```

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.html [hunk 4]
```diff
@@ -1,42 +1,36 @@
...
+    <dot-copy-link
+        *ngIf="field.variable"
+        [label]="field.variable"
+        [copy]="field.variable"
+    ></dot-copy-link>
+    <div class="field__attr__group" *ngIf="fieldAttributes.length">
+        <div class="field__attr" [class.small]="shouldResize && small">
+            <p-chip *ngFor="let field of fieldAttributes" [label]="field"></p-chip>
+        </div>
+        <ng-container *ngIf="shouldResize && small">
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.html:31]: I don't think you need `shouldResize` here because the `small` can validate that in the class

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts [hunk 5]
```diff
@@ -52,18 +66,66 @@ export class ContentTypesFieldDragabbleItemComponent implements OnInit {
...
+
+                    this.cd.detectChanges();
+                } else if (host.contentRect.width > 300 && this.small) {
+                    this.small = false;
+
+                    this.cd.detectChanges();
+                }
+            });
+
+            this.resizeObserver$.observe(this.el.nativeElement);
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts:90]: This is clever! until we can add container queries!

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts [hunk 6]
```diff
@@ -52,18 +66,66 @@ export class ContentTypesFieldDragabbleItemComponent implements OnInit {
...
+        this.open = false;
+    }
+
+    @HostListener('window:click', ['$event'])
+    onWindowClick($event: MouseEvent) {
+        $event.stopPropagation();
+        this.open = false;
+    }
+
+    openAttr($event: MouseEvent) {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts:117]: Add docs.

File: core-web/apps/dotcms-ui/src/app/view/components/dot-api-link/dot-api-link.component.scss [hunk 7]
```diff
@@ -1,35 +1,25 @@
...
-    gap: 0.25rem;
-}
-
-.material-icons {
-    color: $color-palette-primary;
-    font-size: $font-size-default;
-}
-
-// This styles are the same as .p-button.p-button-outlined:enabled:hover in _buttons.scss,
+// This styles are the same as .p-button.p-button-text in _buttons.scss,
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/view/components/dot-api-link/dot-api-link.component.scss:9]: In the future we might want to turn this component into a "link icon" component.

File: core-web/apps/dotcms-ui/src/app/view/components/dot-api-link/dot-api-link.component.html [hunk 8]
```diff
@@ -1,6 +1,6 @@
 <ng-container *ngIf="link">
-    <a class="p-button-sm p-button-outlined" [href]="link" [title]="link" pButton target="_blank">
-        <i class="material-icons">link</i>
-        API
+    <a class="p-button-sm p-button-text" [href]="link" [title]="link" pButton target="_blank">
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/view/components/dot-api-link/dot-api-link.component.html:2]: What the `pButton` directive do to this? doesn't work?

File: core-web/apps/dotcms-ui/src/app/view/components/dot-copy-button/dot-copy-button.component.html [hunk 9]
```diff
@@ -8,6 +8,6 @@
     hideDelay="800"
     tooltipPosition="bottom"
 >
-    <i class="material-icons">content_copy</i>
-    {{ label }}
+    <i class="pi pi-copy"></i>
+    <span class="p-button-label"> {{ showLabel ? label : '' }}</span>
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/view/components/dot-copy-button/dot-copy-button.component.html:12]:
```suggestion
    <span class="p-button-label" *ngIf="!!showLabel">{{ label }}</span>
```

File: core-web/libs/dotcms-scss/shared/_shadows.scss [hunk 10]
```diff
@@ -1,5 +1,9 @@
+@use "./colors" as *;
+
 $md-shadow-1: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
 $md-shadow-2: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
 $md-shadow-3: 0 10px 24px 0 rgba(0, 0, 0, 0.2);
 $md-shadow-4: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
 $md-shadow-5: 0 10px 20px 0 rgba(0, 0, 0, 0.15);
+
+$shadow-focused: 0px 0px 0px 2.8px $color-palette-primary-300;
```
Comment (fmontes) [inline core-web/libs/dotcms-scss/shared/_shadows.scss:9]: Use any of the `$md-shadow` that looks close enough.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/view/components/dot-api-link/dot-api-link.component.html (hunk 8 above)
Comment (zJaaal) [reply core-web/apps/dotcms-ui/src/app/view/components/dot-api-link/dot-api-link.component.html:2]: it just add the primeNG classes

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts (hunk 5 above)
Comment (zJaaal) [reply core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts:90]: I even think about making this a directive, like it receives breakpoints and emits the current breakpoint (small, medium, large)

Breakpoints like: 

```javascript
const breakpoints = {
    small: 100, // Less than 100px is small, emits "small"
    medium: 300
    large: 500
}
```

Something like that

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts
Comment (rjvelazco) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-content-types-edit/components/fields/content-type-field-dragabble-item/content-type-field-dragabble-item.component.ts:1]: Use [document this VS Extension](https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis) to document the function in this file.

File: core-web/apps/dotcms-ui/src/app/view/components/dot-copy-button/dot-copy-button.component.html [hunk 11]
```diff
@@ -8,6 +8,6 @@
     hideDelay="800"
     tooltipPosition="bottom"
 >
-    <i class="material-icons">content_copy</i>
-    {{ label }}
+    <i class="pi pi-copy"></i>
+    <span class="p-button-label" *ngIf="!!showLabel">{{ label }}</span>
```
Comment (rjvelazco) [inline core-web/apps/dotcms-ui/src/app/view/components/dot-copy-button/dot-copy-button.component.html:12]: Isn't the `showLabel` variable already a boolean value?

```suggestion
    <span class="p-button-label" *ngIf="showLabel">{{ label }}</span>
```

File: core-web/libs/dotcms-scss/shared/_shadows.scss [hunk 12]
```diff
@@ -1,3 +1,5 @@
+@use "./colors" as *;
```
Comment (rjvelazco) [inline core-web/libs/dotcms-scss/shared/_shadows.scss:1]: Why is this needed?

**Review by rjvelazco — CHANGES_REQUESTED** [review]

File: core-web/libs/dotcms-scss/shared/_shadows.scss (hunk 12 above)
Comment (zJaaal) [reply core-web/libs/dotcms-scss/shared/_shadows.scss:1]: not needed anymore, I'll delete it

File: core-web/apps/dotcms-ui/src/app/view/components/dot-copy-button/dot-copy-button.component.html (hunk 11 above)
Comment (fmontes) [reply core-web/apps/dotcms-ui/src/app/view/components/dot-copy-button/dot-copy-button.component.html]: Yeah my mistake, I meat `*ngIf="!!label"`
